libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
 
-	FunctionTypePointer functionType;
+	solAssert(!!_functionType, "");
+	solAssert(_functionType->kind() != FunctionType::Kind::ABIDecode, "");
+
+	// Check for unsupported use of bare static call
+	if (
+		_functionType->kind() == FunctionType::Kind::BareStaticCall &&
+		!m_evmVersion.hasStaticCall()
+	)
+		m_errorReporter.typeError(
+			_functionCall.location(),
+			"\"staticcall\" is not supported by the VM version."
+		);
 
-	/// For error message: Struct members that were removed during conversion to memory.
-	set<string> membersRemovedForStructConstructor;
-	if (_functionCall.annotation().kind == FunctionCallKind::StructConstructorCall)
+	// Check for deprecated function names
+	if (_functionType->kind() == FunctionType::Kind::KECCAK256)
 	{
-		TypeType const& t = dynamic_cast<TypeType const&>(*expressionType);
-		auto const& structType = dynamic_cast<StructType const&>(*t.actualType());
-		functionType = structType.constructorType();
-		membersRemovedForStructConstructor = structType.membersMissingInMemory();
-		_functionCall.annotation().isPure = isPure;
+		if (auto functionName = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+			if (functionName->name() == "sha3")
+				m_errorReporter.typeError(
+					_functionCall.location(),
+					"\"sha3\" has been deprecated in favour of \"keccak256\""
+				);
 	}
-	else if ((functionType = dynamic_pointer_cast<FunctionType const>(expressionType)))
-		_functionCall.annotation().isPure =
-			isPure &&
-			_functionCall.expression().annotation().isPure &&
-			functionType->isPure();
-
-	bool allowDynamicTypes = m_evmVersion.supportsReturndata();
-	if (!functionType)
+	else if (_functionType->kind() == FunctionType::Kind::Selfdestruct)
 	{
-		m_errorReporter.typeError(_functionCall.location(), "Type is not callable");
-		_functionCall.annotation().type = make_shared<TupleType>();
-		return false;
+		if (auto functionName = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+			if (functionName->name() == "suicide")
+				m_errorReporter.typeError(
+					_functionCall.location(),
+					"\"suicide\" has been deprecated in favour of \"selfdestruct\""
+				);
 	}
 
-	if (functionType->kind() == FunctionType::Kind::BareStaticCall && !m_evmVersion.hasStaticCall())
-		m_errorReporter.typeError(_functionCall.location(), "\"staticcall\" is not supported by the VM version.");
+	// Check for event outside of emit statement
+	if (!m_insideEmitStatement && _functionType->kind() == FunctionType::Kind::Event)
+		m_errorReporter.typeError(
+			_functionCall.location(),
+			"Event invocations have to be prefixed by \"emit\"."
+		);
+
+	// Perform standard function call type checking
+	typeCheckFunctionGeneralChecks(_functionCall, _functionType);
+}
+
+void TypeChecker::typeCheckABIEncodeFunctions(
+	FunctionCall const& _functionCall,
+	FunctionTypePointer _functionType
+)
+{
+	solAssert(!!_functionType, "");
+	solAssert(
+		_functionType->kind() != FunctionType::Kind::ABIEncode ||
```

Comment:
These should be equalities, not inequalities, right? This way the assertion is always true and checks nothing.

Reply:
Yes good catch!